File: src/Error1.py
# ---------------------------------------------------------------------------
# File    : error1.py
# Purpose : Compute standard error/skill metrics for hydrological series and plot.
# Author  : Gerald Augusto Corzo Pérez
# Affil.  : IHE Delft — Hydroinformatics
# Version : 2.1 (2025-08-31)
# License : MIT
# SPDX-License-Identifier: MIT
#
# Metrics returned (dict):
#   RMSE, NSC, Cor, NRMSE (%), MAE, StdT, StdP, MuT, MuP,
#   PERS, SSE, SSEN, RMSEN, NRMSEN (%), MARE, Er (array), Po, Pu, fig (matplotlib Figure or None)
#
# Conventions:
#   - Pairwise drop NaNs (like MATLAB nansum/nanmean usage in your script).
#   - NRMSE and NRMSEN are percentages of population std (ddof=0) to match `std(x,1)` in MATLAB.
#   - Coefficient of persistence: PERS = 1 - SSE / SSEN, where SSEN is the naive (persistence) SSE.
# ---------------------------------------------------------------------------

from __future__ import annotations
from typing import Dict, Any, Optional
import numpy as np

def error1(P, T, comment: str = "Target and Predicted", plot: bool = True) -> Dict[str, Any]:
    import matplotlib.pyplot as plt  # local import so the module works without plotting deps if needed

    P = np.asarray(P, dtype=float).ravel()
    T = np.asarray(T, dtype=float).ravel()
    if P.shape != T.shape:
        raise ValueError("P and T must have the same length.")
    mask = np.isfinite(P) & np.isfinite(T)
    P, T = P[mask], T[mask]

    n = P.size
    if n < 2:
        raise ValueError("Need at least 2 paired values after removing NaNs/Infs.")

    # Means & population std (ddof=0) to mirror MATLAB std(x,1)
    MuT = float(np.mean(T))
    MuP = float(np.mean(P))
    StdT = float(np.std(T, ddof=0))
    StdP = float(np.std(P, ddof=0))

    # Errors
    resid = P - T
    SSE  = float(np.sum(resid**2))
    RMSE = float(np.sqrt(SSE / n))
    MAE  = float(np.mean(np.abs(resid)))

    # NRMSE as % of Std(T)
    NRMSE = float(100.0 * RMSE / StdT) if StdT > 0 else np.inf

    # Nash–Sutcliffe (NSC/NSE)
    denom_nsc = float(np.sum((T - MuT) ** 2))
    NSC = float(1.0 - SSE / denom_nsc) if denom_nsc > 0 else np.nan

    # Pearson correlation (explicit formula to mirror MATLAB)
    num_cor = float(np.sum((P - MuP) * (T - MuT)))
    den_cor = float(np.sqrt(np.sum((P - MuP) ** 2) * np.sum((T - MuT) ** 2)))
    Cor = num_cor / den_cor if den_cor > 0 else np.nan

    # MARE: exclude zeros in T
    nz = np.abs(T) > 0
    MARE = float(np.mean(np.abs((T[nz] - P[nz]) / T[nz]))) if np.any(nz) else np.nan

    # Persistence baseline: T_hat[i] = T[i-1]
    T1 = T[:-1]
    T2 = T[1:]
    dT = T1 - T2
    SSEN = float(np.sum(dT**2))
    RMSEN = float(np.sqrt(SSEN / max(1, T2.size)))          # (n-1) terms
    stdT2 = float(np.std(T2, ddof=0))                       # match MATLAB std(T2,1)
    NRMSEN = float(100.0 * RMSEN / stdT2) if stdT2 > 0 else np.inf

    PERS = float(1.0 - SSE / SSEN) if SSEN > 0 else np.nan

    Er = T - P
    Po = float(np.sum(Er <= 0) / Er.size)  # proportion <= 0
    Pu = float(np.sum(Er >  0) / Er.size)  # proportion  > 0

    fig: Optional["plt.Figure"] = None
    if plot:
        fig = plt.figure(figsize=(8, 6), layout="constrained")
        fig.canvas.manager.set_window_title(comment) if hasattr(fig.canvas.manager, "set_window_title") else None
        import matplotlib.pyplot as plt_
        ax1 = fig.add_subplot(2, 1, 1)
        ax1.plot(P, "r--", linewidth=1.8, label="Predicted")
        ax1.plot(T, "b-",  linewidth=1.2, label="Target")
        ax1.legend(loc="lower right")
        ax1.set_title(f"Time series of target and predicted\n(RMSE = {RMSE:.4g})  {comment}")
        ax1.set_xlabel("Time"); ax1.set_ylabel(r"Discharge (m$^3$/s)")
        ax1.grid(True)

        ax2 = fig.add_subplot(2, 1, 2)
        ax2.stem(np.arange(Er.size), Er, use_line_collection=True)
        ax2.set_title(f"Error between target and predicted (NRMSE = {NRMSE:.2f}%)")
        ax2.set_xlabel("Time"); ax2.set_ylabel("Error (T - P)")
        ax2.grid(True)

    return {
        "RMSE": RMSE, "NSC": NSC, "Cor": Cor, "NRMSE": NRMSE, "MAE": MAE,
        "StdT": StdT, "StdP": StdP, "MuT": MuT, "MuP": MuP,
        "PERS": PERS, "SSE": SSE, "SSEN": SSEN, "RMSEN": RMSEN, "NRMSEN": NRMSEN,
        "MARE": MARE, "Er": Er, "Po": Po, "Pu": Pu, "fig": fig
    }

File: matlab/Error1.m
function Error = Error1(P, T, Comment)
% -------------------------------------------------------------------------
% File    : Error1.m
% Purpose : Compute standard error/skill metrics for hydrological series and plot.
% Author  : Gerald Augusto Corzo Pérez
% Affil.  : IHE Delft — Hydroinformatics
% Version : 2.1 (2025-08-31)  % cleaned NaN handling and plotting
% License : MIT
% SPDX-License-Identifier: MIT
%
% Usage
%   Error = Error1(P, T, Comment)
% Inputs
%   P : Predicted values (vector)
%   T : Target/Observed values (vector)
%   Comment (optional) : figure tag/title suffix
%
% Outputs (fields)
%   RMSE, NSC, Cor, NRMSE (%), MAE, StdT, StdP, MuT, MuP,
%   PERS, SSE, SSEN, RMSEN, NRMSEN (%), MARE, Er, Po, Pu, h
%
% Notes
%   - NRMSE = 100 * RMSE / Std(T)     % your convention
%   - To use range-based NRMSE instead, set rngT = max(T)-min(T) and use:
%       NRMSE = 100 * RMSE / rngT;
%   - Coefficient of persistence PERS = 1 - SSE / SSEN (persistence baseline)
% -------------------------------------------------------------------------

if nargin < 3 || isempty(Comment)
    Comment = 'Target and Predicted';
end

% Ensure column vectors and pairwise finite values only
P = P(:); T = T(:);
mask = isfinite(P) & isfinite(T);
P = P(mask); T = T(mask);

n = numel(P);
if n < 2
    error('Error1:NotEnoughData', 'Need at least 2 paired values.');
end

% Basic stats (MATLAB-style population std uses w=1; omit NaNs already)
MuT  = mean(T, 'omitnan');   MuP  = mean(P, 'omitnan');
StdT = std(T, 1, 'omitnan'); StdP = std(P, 1, 'omitnan');

% Errors
resid = P - T;
SSE  = sum(resid.^2, 'omitnan');
RMSE = sqrt(SSE / n);
MAE  = mean(abs(resid), 'omitnan');

% NRMSE (% of Std(T)) -- change here if you prefer range-based
NRMSE = 100 * RMSE / StdT;    % if StdT==0 -> Inf; keep as-is to signal degenerate series

% Nash–Sutcliffe (NSC/NSE)
den = sum((T - MuT).^2, 'omitnan');
NSC = 1 - SSE / den;

% Pearson correlation (guard zero variance)
if StdT > 0 && StdP > 0
    Cor = corr(P, T);
else
    Cor = NaN;
end

% MARE (exclude zeros in T to avoid division by zero)
nz = (T ~= 0);
if any(nz)
    MARE = mean(abs((T(nz) - P(nz)) ./ T(nz)), 'omitnan');
else
    MARE = NaN;
end

% Persistence baseline: T_hat(i) = T(i-1)
T1 = T(1:end-1);
T2 = T(2:end);
m2 = isfinite(T1) & isfinite(T2);
dT = T2(m2) - T1(m2);

SSEN  = sum(dT.^2, 'omitnan');
RMSEN = sqrt(SSEN / max(1, numel(dT)));
% Range- or std-based normalization; here keep % of Std(T)
NRMSEN = 100 * RMSEN / StdT;

% Pack struct
Error.RMSE  = RMSE;
Error.NSC   = NSC;
Error.Cor   = Cor;
Error.NRMSE = NRMSE;
Error.MAE   = MAE;
Error.StdT  = StdT;
Error.StdP  = StdP;
Error.MuT   = MuT;
Error.MuP   = MuP;
Error.PERS  = 1 - SSE / SSEN;   % coefficient of persistence
Error.SSE   = SSE;
Error.SSEN  = SSEN;             % SSE of naive (persistence)
Error.RMSEN = RMSEN;
Error.NRMSEN= NRMSEN;
Error.MARE  = MARE;

% Error sign stats
Error.Er    = T - P;
Error.Po    = sum(Error.Er <= 0) / numel(Error.Er);  % proportion <= 0
Error.Pu    = sum(Error.Er >  0) / numel(Error.Er);  % proportion >  0

% Plot
figure1 = figure('Color',[1 1 1], 'Tag', Comment, 'HandleVisibility','on','Visible','on');
tiledlayout(2,1, 'Padding','compact', 'TileSpacing','compact');

% Upper: series
nexttile;
H1 = plot(P,'r--','LineWidth',1.8); hold on;
H2 = plot(T,'b-','LineWidth',1.2);
legend([H1 H2], {'Predicted','Target'}, 'Location','southoutside');
title({ 'Time series of target and predicted', ...
        sprintf('(RMSE = %.4g)  %s', RMSE, Comment) });
xlabel('Time'); ylabel('Discharge (m^3/s)'); grid on;

% Lower: error (T - P)
nexttile;
stem(T - P, 'filled'); grid on;
title(sprintf('Error between target and predicted (NRMSE = %.2f%%)', NRMSE));
xlabel('Time'); ylabel('Error (T - P)');

Error.h = figure1;
end
